[Backend] Clean up MMAv3 part of MatmulLoopPipeline.

jlebar opened the pull request:
<git-pr-chain>


[Backend] Clean up MMAv3 part of MatmulLoopPipeline.

The MMAv3 pipelining code needed some love and comments.

There are some minor functional changes in here.  For example, the old
code relied on a particular order of op.users(), and I'm not convinced
it was correct.

Once you unravel the code, what it needs to do is a lot less complicated
than what it was doing.  Indeed, despite the fact that I added a ton of
comments, it's roughly the same length as before.

The changes to the test are cosmetic.  Previously we emitted tt.dot and
assumed it would eventually becdome ttng.dot_async in a later pass (or
at least during codegen).  Now we explicitly convert some tt.dot's to
ttng.dot_async so we can have an explicit wait and thread the results
through.

There's more cleanup that can be done here (see the TODOs added in this
PR), but I needed to stop somewhere.


#### [PR chain](https://github.com/jlebar/git-pr-chain)
1. 👉 #3301 👈 **YOU ARE HERE**


</git-pr-chain>

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+  // Rule 1: All shmem operands are used "safely".
+  auto checkOperand = [&](Value operand) {
+    // If it's a shmem operand, it must either be defined outside the loop, or
+    // come from an ExtractSlice op.  Only ConvertLayout and Trans ops are
```
replace `ExtractSlice` with `MemDescSubview`

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+  if (!dotOp)
+    return std::nullopt;
+
+  // Rule 1: All shmem operands are used "safely".
```
nit: I would replace used "safely" by "have been multi-buffered"

ThomasRaoux replied:
not sure if you missed that one or prefer to leave it as is, commenting just in case

jlebar replied:
Thanks, fixed.

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
-          }
-        }
-      }
+  // If we have a synchyronizing dot, convert it to async and insert a wait
+  // right after it.  Otherwise, we'll insert the wait at the end of the loop.
+  IRRewriter builder(forOp.getContext());
+  if (synchronizingDot) {
+    builder.setInsertionPointAfter(synchronizingDot);
+    builder.replaceOpWithNewOp<ttng::DotAsyncOp>(
+        synchronizingDot.getOperation(), synchronizingDot.getA(),
+        synchronizingDot.getB(), synchronizingDot.getC(),
+        synchronizingDot.getAllowTF32(),
+        synchronizingDot.getMaxNumImpreciseAcc());
+  } else {
+    builder.setInsertionPoint(forOp.getBody()->getTerminator());
+  }
+
+  SmallVector<Value> waitOperands;
+  for (auto [asyncDot, iterArgIdx] : asyncDots) {
+    Value dotResult = syncAtBeginning
+                          ? cast<Value>(forOp.getRegionIterArg(iterArgIdx))
+                          : cast<Value>(asyncDot->getResult(0));
+    if (!dotResult.use_empty())
+      waitOperands.push_back(dotResult);
+  }
```
Is inserting the wait at the end of the loop correct? That wasn't the behavior of previous code right?

jlebar replied:
Great point, fixed.

ThomasRaoux reviewed (commented):
added few nits but I'm also not the logic is fully correct based on the comments (or maybe there is a mismatch between the comment and the code)

jlebar commented on test/TritonGPU/loop-pipeline-hopper.mlir:
```diff
@@ -384,7 +384,7 @@ module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-c
     // CHECK: scf.for
     // CHECK:   triton_gpu.async_wait {{.*}} {num = 1 : i32}
-    // CHECK:   tt.dot
+    // CHECK:   triton_nvidia_gpu.dot_async
```
Right; added a check.

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -129,61 +130,59 @@ createAsyncCopy(scf::ForOp &forOp, tt::LoadOp loadOp, Value alloc,
+  return attr;
 }
 
+// TODO: This returns true and *sometimes* sets enc?
```
I think I'm pretty likely to get to them, yeah.

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+    return std::nullopt;
+
+  // Rule 2: The dot should only be used by the for loop's `yield`.
+  if (!dotOp->hasOneUse() || !isa<scf::YieldOp>(*dotOp->getUsers().begin()))
```
Great catch.

ThomasRaoux commented on test/TritonGPU/pipeline-hopper-remove-wait.mlir:
```diff
@@ -114,6 +114,10 @@ module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-c
+      // The first dot gets converted to dot-async + wait.  The second one
+      // doesn't have a wait because the first wait is sufficient.
+      // CHECK: triton_nvidia_gpu.dot_async
+      // CHECK: triton_nvidia_gpu.dot_wait
```
also maybe changing it to a CHECK-NEXT would ensure the wait is at the right spot

jlebar replied:
Done

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+// the dependencies through the wait op.  This keeps other passes from
+// reordering the dots.
+//
+// TODO(jlebar): It's possible that the sync dot might appear in the wrong place
+// but could be hoisted/sunk to the right place.  Right now we don't handle
+// this.
```
I was thinking of something like this.

%a = dot.async %x
%b = dot %y
%c = dot.async %b

The second dot has to be synchronous because it's used by the third dot.  But the first two dots do not depend on each other, so we can do them in any order.  It would therefore be better to do %b first and then do %a -- that way we could run %a and %c at the same time.

Does that not work?

jlebar replied:
I don't see how you'd do it beforehand, because until this pass runs, you don't know which dots are going to be sync vs async?  And if you did it afterwards, you'd have to repeat all of the work this pass does to determine if it's safe to move?

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
-    while (op->getBlock()->getParentOp() != forOp) {
-      lastOp = op;
-      op = op->getBlock()->getParentOp();
+// If necessary, insert a dot-wait inside the loop, waiting for the results of
+// the async dots from iteration i-2 to complete.  (We pipeline to depth 2, so
+// there are at most 2 dots in flight at any point in time.)
```
Thanks, fixed.

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
   }
 
-  hasDotWait0 = hasDotWait0 || hasSyncDot;
+  // Next, insert a wait inside the loop.  We pipeline to depth 2, so the
+  // third iteration's set of asynchronous dots can't start until the first
+  // iteration's set has completed.
```
I think this one is correct?  This says that iteration i+2 cannot *begin* until iteration i completes, iow we have 2 iterations in flight.  Whereas the other comment was saying that iteration i waits on iteration i-2 after issuing its loads, i.e. we'd end up with *three* iterations in flight.

ThomasRaoux replied:
The logic is more that we cannot start the copies for iteration i until the wgmma for iteration i-1 is done. The problem is bot as much about the number of wgmma flight than how many shared slices it is locking. I wonder if we could make it more clear from the comment.

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -129,61 +130,59 @@ createAsyncCopy(scf::ForOp &forOp, tt::LoadOp loadOp, Value alloc,
-allTransitiveUsesHaveDotEncoding(Value val,
-                                 ttg::SharedEncodingAttr &sharedEnc) {
+//
+// TODO: Rename
```
clarified the todo

jlebar commented:
Thank you for the review, Thomas.  I've pushed a change that addresses your comments.  (Wish there was a way to wait to reply to the comments until after I pushed the change, but I don't see a good way.)

ThomasRaoux commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +872,242 @@ void mlir::triton::updateWaits(ModuleOp module) {
+  };
+
+  if (!checkOperand(dotOp.getA()) || !checkOperand(dotOp.getB()) ||
+      !checkOperand(dotOp.getC()))
```
technically we only care about A and B operand. Are you adding C for completeness? I guess it is always a MMA reg

jlebar replied:
Good point, fixed.

ThomasRaoux reviewed (commented):
few more minor comments about the algo description

jlebar commented on lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp:
```diff
@@ -856,199 +870,238 @@ void mlir::triton::updateWaits(ModuleOp module) {
+// the async dots from iteration i-2 to complete.  (We pipeline to depth 2, so
+// there are at most 2 dots in flight at any point in time.)
+//
+// TODO(jlebar): I'm told CUTLASS pipelines to depth 4 or even 8, to better hide
+// variable-latency operations.  Would this help us?
```
Sure

jlebar commented:
Thanks again for the careful review.  In addition to addressing your comments, I made some minor changes to how waits are inserted.  Previously (matching the original code) we only inserted the waits if the dots were used.  But actually I don't think this is totally right.  First of all, DCE should have gotten rid of unused dots.  But if for some reason one survives, we still should wait for it to complete, because that shmem buffer might be needed by something else.

ThomasRaoux commented:
> I made some minor changes to how waits are inserted. Previously (matching the original code) we only inserted the waits if the dots were used. But actually I don't think this is totally right. First of all, DCE should have gotten rid of unused dots. But if for some reason one survives, we still should wait for it to complete, because that shmem buffer might be needed by something else.

maybe we should just remove dot with empty use on the fly as I don't think we want special logic for this case. 
I'm not sure I understand why we would need to have a wait for other shared memory usage, if the dot with no users reads random data out of shared memory this should not be observable.

ThomasRaoux reviewed (approved):
LGTM

jlebar commented:
> maybe we should just remove dot with empty use on the fly as I don't think we want special logic for this case.

Quite the opposite, there's now *less* special logic.  We used to have special logic to handle a dot with no users, but I think this should be the responsibility of DCE to handle.  If we have a dot by the time it gets to this pass, we should just treat it like we treat everything else.

ThomasRaoux commented:
> > maybe we should just remove dot with empty use on the fly as I don't think we want special logic for this case.
> 
> Quite the opposite, there's now _less_ special logic. We used to have special logic to handle a dot with no users, but I think this should be the responsibility of DCE to handle. If we have a dot by the time it gets to this pass, we should just treat it like we treat everything else.

yes makes  sense
